src/core/geometry/qgsnurbscurve.cpp

```diff
@@ -0,0 +1,1622 @@
+}
+
+QgsLineString *QgsNurbsCurve::curveToLine( double tolerance, SegmentationToleranceType toleranceType ) const
+{
+  Q_UNUSED( toleranceType );
+
+  // Determine number of segments based on tolerance (angular approximation)
+  // For NURBS curves, we use uniform parameterization as a first approximation
+  const int steps = std::max( 2, static_cast<int>( 2 * M_PI / tolerance ) );
+
+  auto line = new QgsLineString();
+  for ( int i = 0; i <= steps; ++i )
+  {
+    const double t = static_cast<double>( i ) / steps;
+    const QgsPoint pt = evaluate( t );
+    line->addVertex( pt );
+  }
+
+  return line;
+}
+
+void QgsNurbsCurve::draw( QPainter &p ) const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  if ( line )
+    line->draw( p );
+}
+
+void QgsNurbsCurve::drawAsPolygon( QPainter &p ) const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  if ( line )
+    line->drawAsPolygon( p );
+}
+
+QPolygonF QgsNurbsCurve::asQPolygonF() const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  return line ? line->asQPolygonF() : QPolygonF();
+}
+
+QgsPoint QgsNurbsCurve::endPoint() const
+{
+  return mControlPoints.isEmpty() ? QgsPoint() : mControlPoints.last();
+}
+
+bool QgsNurbsCurve::equals( const QgsCurve &other ) const
+{
+  if ( geometryType() != other.geometryType() )
+  {
+    return false;
+  }
+
+  const QgsNurbsCurve *o = qgsgeometry_cast<const QgsNurbsCurve *>( &other );
+  if ( !o )
+    return false;
+
+  if ( o->mDegree != mDegree || o->mClosed != mClosed || o->mControlPoints.size() != mControlPoints.size() )
+  {
+    return false;
+  }
+
+  if ( o->mWeights.size() != mWeights.size() || o->mKnots.size() != mKnots.size() )
+  {
+    return false;
+  }
+
+  for ( int i = 0; i < mControlPoints.size(); ++i )
+  {
+    if ( mControlPoints[i] != o->mControlPoints[i] )
+      return false;
+    if ( i < mWeights.size() && i < o->mWeights.size() )
+    {
+      if ( !qgsDoubleNear( mWeights[i], o->mWeights[i] ) )
+        return false;
+    }
+  }
+
+  for ( int i = 0; i < mKnots.size(); ++i )
+  {
+    if ( !qgsDoubleNear( mKnots[i], o->mKnots[i] ) )
+      return false;
+  }
+
+  return true;
+}
+
+int QgsNurbsCurve::indexOf( const QgsPoint &point ) const
+{
+  for ( int i = 0; i < mControlPoints.size(); ++i )
+  {
+    if ( qgsDoubleNear( mControlPoints[i].distance( point ), 0.0 ) )
+    {
+      return i;
+    }
+  }
+  return -1;
+}
+
+QgsPoint *QgsNurbsCurve::interpolatePoint( double distance ) const
+{
+  std::unique_ptr<QgsLineString> line( curveToLine() );
+  if ( !line )
+  {
+    return nullptr;
+  }
+  return line->interpolatePoint( distance );
+}
+
+int QgsNurbsCurve::numPoints() const
+{
+  return mControlPoints.size();
+}
+
+bool QgsNurbsCurve::pointAt( int node, QgsPoint &point, Qgis::VertexType &type ) const
+{
+  if ( node < 0 || node >= mControlPoints.size() )
+  {
+    return false;
+  }
+  point = mControlPoints[node];
+  type = Qgis::VertexType::Segment;
```

Comment:
is Qgis::VertexType::Segment relevant ? The doc says "The actual start or end point of a segment". Looks to me like Curve would be more appropriate. But that might be also dependent on the type of the NURBS if the geometry actually passes by the control point? Or maybe add a new value to the VertexType enumeration ?

Reply:
I added Qgis::VertexType::ControlPoint